### PR TITLE
Optimize checksum creation for remote cluster state

### PR DESCRIPTION
**Pull request by himshikha**
### Description
This change parallelizes checksum creation for cluster state components, with the aim to reduce remote state publication latency.

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by Bukhtawar — COMMENTED** [review]
Thanks @himshikha would you be able to share the perf numbers with this change

File: server/src/main/java/org/opensearch/gateway/remote/ClusterStateChecksum.java [hunk 1]
```diff
@@ -66,61 +73,143 @@ public class ClusterStateChecksum implements ToXContentFragment, Writeable {
...
+        }, checksum -> clusterStateCustomsChecksum = checksum, executorService, latch);
+
+        executorService.shutdown();
+        try {
+            latch.await();
+        } catch (InterruptedException e) {
+            throw new RemoteStateTransferException("Failed to create checksum for cluster state.", e);
+        }
+        createClusterStateChecksum();
+        logger.debug("Checksum execution time {}", System.currentTimeMillis() - start);
```
[the remark this answers is not in the document]
Comment (himshikha) [reply server/src/main/java/org/opensearch/gateway/remote/ClusterStateChecksum.java]: done

**Comment by himshikha** [comment]
> ❌ Gradle check result for [80d0890](https://github.com/opensearch-project/OpenSearch/commit/80d0890952515f9b4c29a306a30858d65955a1ea): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/48382/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky test https://github.com/opensearch-project/OpenSearch/issues/14294, https://github.com/opensearch-project/OpenSearch/issues/16070, https://github.com/opensearch-project/OpenSearch/issues/14302, https://github.com/opensearch-project/OpenSearch/issues/15814, https://github.com/opensearch-project/OpenSearch/issues/15844

**Comment by himshikha** [comment]
> ❌ Gradle check result for [c0321b1](https://github.com/opensearch-project/OpenSearch/commit/c0321b1a11d00055f54ee8fcf2f57127531edf17): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/48506/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky test https://github.com/opensearch-project/OpenSearch/issues/15812

File: server/src/main/java/org/opensearch/gateway/remote/ClusterStateChecksum.java [hunk 2]
```diff
@@ -65,62 +72,147 @@ public class ClusterStateChecksum implements ToXContentFragment, Writeable {
...
+
+        executeChecksumTask((stream) -> {
+            try {
+                clusterState.nodes().writeVerifiableTo(stream);
+            } catch (IOException e) {
+                throw new RemoteStateTransferException("Failed to create checksum for discovery nodes", e);
+            }
+            return null;
+        }, checksum -> nodesChecksum = checksum, executorService, latch);
+
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/gateway/remote/ClusterStateChecksum.java:97]: Can we take a Consumer<Stream> as ` clusterState.nodes().writeVerifiableTo(stream);` and make the entire block re-usable

File: server/src/main/java/org/opensearch/gateway/remote/ClusterStateChecksum.java [hunk 3]
```diff
@@ -65,62 +73,103 @@ public class ClusterStateChecksum implements ToXContentFragment, Writeable {
...
+        executorService.execute(() -> {
+            try {
+                long checksum = createChecksum(checksumTask);
+                checksumConsumer.accept(checksum);
+                latch.countDown();
+            } catch (IOException e) {
+                throw new RemoteStateTransferException("Failed to execute checksum task", e);
+            }
+        });
+    }
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/gateway/remote/ClusterStateChecksum.java:163]: Would we be able to release latches in the exception paths?

Comment (himshikha) [reply server/src/main/java/org/opensearch/gateway/remote/ClusterStateChecksum.java:162]: Checksum creation will fail on exception. Latch should automatically release, tested exception cases. But I can move it to finally block.

**Comment by himshikha** [comment]
> ❌ Gradle check result for [93f9a43](https://github.com/opensearch-project/OpenSearch/commit/93f9a4339ae23af991297b6239c9c1b21521f22c): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/48615/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky test https://github.com/opensearch-project/OpenSearch/issues/14324

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java [hunk 4]
```diff
@@ -191,6 +192,7 @@ public static ThreadPoolType fromType(String type) {
         map.put(Names.REMOTE_RECOVERY, ThreadPoolType.SCALING);
         map.put(Names.REMOTE_STATE_READ, ThreadPoolType.SCALING);
         map.put(Names.INDEX_SEARCHER, ThreadPoolType.RESIZABLE);
+        map.put(Names.REMOTE_STATE_CHECKSUM, ThreadPoolType.SCALING);
```
[the remark this answers is not in the document]
Comment (himshikha) [reply server/src/main/java/org/opensearch/threadpool/ThreadPool.java]: Changed to Fixed pool with thread size equal to number of components in cluster state

**Comment by himshikha** [comment]
> ❌ Gradle check result for [f14898c](https://github.com/opensearch-project/OpenSearch/commit/f14898c81a3896119a75b8e181517fb65c800da8): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/48705/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky test https://github.com/opensearch-project/OpenSearch/issues/15944, https://github.com/opensearch-project/OpenSearch/issues/15812

**Review by Bukhtawar — APPROVED** [review]

**Comment by sandeshkr419** [comment]
Thanks @himshikha for this change.

@Bukhtawar @himshikha - I think changes like this should have a `changelog` so that they can be called out in performance gains / optimizations as part of release notes.

**Pull request merged**